Single inheritance

kunjsong01 opened the pull request:
https://github.com/esbmc/esbmc/pull/854 is too large and messy for review. 

Based on previous discussion on slack, this is a PR separated from https://github.com/esbmc/esbmc/pull/854, adding support for single inheritance only. 

Features implemented: 
1. Populated class type's `components` and `methods` vectors (which will be used for polymorphism support, e.g. generating implicit ctor/dtor code for vptr initializations).
2. Added base initializer conversion and generate a type casted argument for base initializer.
3. Added the corresponding test cases.

kunjsong01 committed:
[utli] added debug lines in show_symbol_table

kunjsong01 committed:
[clang-cpp-converter] bypass errors to generate symbol table

kunjsong01 committed:
[clang-cpp-converter] added #class and name for struct type

kunjsong01 committed:
[clang-cpp-converter] added tag_prefix

kunjsong01 committed:
[clang-c*-converter] refactored get_struct_union_class_fields

kunjsong01 committed:
[clang-c*-converter] added location for class/struct/union component

kunjsong01 committed:
[clang-cpp-converter] added member_name to indicate parent for class field

kunjsong01 committed:
[clang-cpp-converter] added access in class component's type

kunjsong01 committed:
[clang-c-converter] added location in class symbol type

kunjsong01 committed:
[clang-c*-converter] synced component type with function type for class methods

kunjsong01 committed:
[clan-c*-converter] added member_name in component type for class method

kunjsong01 committed:
[clang-c-converter] fixed linkage

kunjsong01 committed:
[clang-c-converter] fixed inline

kunjsong01 committed:
[clang-cpp-converter] additional annotations for class method component

kunjsong01 committed:
[clang-c*-converter] fixed ctor argument in derived class

kunjsong01 committed:
[irep][clang-cpp-converter] refactored implicit this for base ctor

kunjsong01 committed:
[regression] added test cases

kunjsong01 committed:
[clang-cpp-converter] removed debugging lines

kunjsong01 commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -1024,3 +1097,138 @@ bool clang_cpp_convertert::get_decl_ref(
+{
+  // sanity checks
+  assert(initializer.derived_this_arg() != "");
+  assert(initializer.base_ctor_derived());
```
@mikhailramalho 
This function would appear to be adjustment only. 
Shall we move it to adjuster?

mikhailramalho replied:
It's hard to tell... can you explain why we need this function?

kunjsong01 replied:
>can you explain why we need this function?

to generate typecasted `this` argument when calling base ctor in derived ctor. 
Please refer to the comment https://github.com/esbmc/esbmc/pull/882#discussion_r1088164013

lucasccordeiro commented on regression/esbmc-cpp/inheritance_bringup/single_member_fail/main.cpp:
```diff
@@ -0,0 +1,25 @@
+{
+public:
+  Motorcycle() : Vehicle() {}
+  //Motorcycle() {}
```
Remove this line.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -3291,3 +3280,51 @@ clang_c_convertert::get_top_FunctionDecl_from_Stmt(const clang::Stmt &stmt)
   return nullptr;
 }
+
+bool clang_c_convertert::check_alignment_attributes(
```
You should comment on this method.

lucasccordeiro commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -3291,3 +3280,51 @@ clang_c_convertert::get_top_FunctionDecl_from_Stmt(const clang::Stmt &stmt)
+    const auto &attrs = field->getAttrs();
+    for(const auto &attr : attrs)
+    {
+      if(attr->getKind() == clang::attr::Aligned)
```
Add explanation.

lucasccordeiro commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -220,7 +221,19 @@ bool clang_cpp_convertert::get_function(
     return false;
 
-  return clang_c_convertert::get_function(fd, new_expr);
+  // Only convert non-trivial functions, e.g. non-trivial ctor/dtor, copy ctor, move ctor .etc
```
Remove this block of comments.

lucasccordeiro commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -220,7 +221,19 @@ bool clang_cpp_convertert::get_function(
+    return true;
+
+  // add additional annotations for class/struct/union methods
+  if(const auto *md = llvm::dyn_cast<clang::CXXMethodDecl>(&fd))
```
Shall we write it as an AND statement?

kunjsong01 replied:
It checks to see if the operand is of the specified type, and if so, returns a pointer to it, otherwise returns a nullptr.  

This is part of the `clang_cpp_convertert::get_function` which is used to convert both ordinary functions and class methods. So we might get an ordinary function rather than the non-CXXMethodDecl for class method, in which case we'll get a nullptr.

lucasccordeiro commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -268,23 +282,60 @@ bool clang_cpp_convertert::get_struct_union_class_fields(
+    {
+      // `base` type is clang::CXXBaseSpecifier
+      base.getType().dump();
+      //log_error("inheritance is not supported in {}", __func__);
```
Shall we remove these lines?

kunjsong01 replied:
Removed, this will be part of the next PR "split-polymorphism".

lucasccordeiro commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -321,6 +372,10 @@ bool clang_cpp_convertert::get_struct_union_class_methods(
     {
       // Add only if it isn't static
       if(!cxxmd->isStatic())
+      {
```
Remove brackets?

lucasccordeiro reviewed (changes requested):
Some minor comments.

kunjsong01 committed:
[clang-cpp-converter] fixed comments

lucasccordeiro reviewed (approved):
LGTM.

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -298,13 +303,22 @@ bool clang_c_convertert::get_struct_union_class(const clang::RecordDecl &rd)
     t = struct_typet();
   t.tag(name);
 
+  if(rd.isClass())
```
is this needed?

kunjsong01 replied:
I suppose so... just need to distinguish class from struct, making it more intuitive in the symbol table, otherwise it would mark everything `struct`:
![Screenshot 2023-01-25 at 12 51 00](https://user-images.githubusercontent.com/32592800/214568146-ba4906d5-eeaa-4523-b136-e015dcbfc45f.png)

mikhailramalho replied:
If it's just to print "class" instead of "struct", we don't really need it.

kunjsong01 replied:
done removed

mikhailramalho commented on regression/esbmc-cpp/inheritance_bringup/single_member/main.cpp:
```diff
@@ -0,0 +1,25 @@
+{
+public:
+  Motorcycle() : Vehicle() {}
+  //Motorcycle() {}
```
remove

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -433,7 +418,7 @@ bool clang_c_convertert::get_struct_union_class_fields(
 bool clang_c_convertert::get_struct_union_class_methods(
   const clang::RecordDecl &,
-  struct_union_typet &)
+  struct_typet &)
```
why this change? I think unions can have methods.

kunjsong01 replied:
> why this change?

Need to use the `methods()` as in `type.methods().push_back(comp);`. Our `struct_union_typet` class doesn't have this function.

mikhailramalho replied:
Do we need to add the methods to the type? i.e., is there anywhere in esbmc after the frontend (goto, ssa, smt), that we call methods() to get the list of methods?

If it's only to add the methods to the type, I'd say we don't need it.

kunjsong01 replied:
Yes, to generate the implicit code for constructors for vptr initializations.

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -3038,7 +3027,7 @@ std::string clang_c_convertert::get_decl_name(const clang::NamedDecl &nd)
 }
 
-static std::string
+std::string
```
this can be an inline static function, why you removed it?

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -3291,3 +3280,51 @@ clang_c_convertert::get_top_FunctionDecl_from_Stmt(const clang::Stmt &stmt)
+  return false;
+}
+
+bool clang_c_convertert::is_field_global_storage(const clang::FieldDecl *field)
```
do we really need a separate function for this?

mikhailramalho replied:
I see why you need it now: to reuse in the cpp convert. Please use it in clang_cpp_convertert::get_struct_union_class_fields, in the loop after `  // let's check for (virtual) base classes if this declaration is a CXXRecordDecl,`

kunjsong01 replied:
>Please use it in clang_cpp_convertert::get_struct_union_class_fields, in the loop after   // let's check for (virtual) base classes if this declaration is a CXXRecordDecl,

Yep, it's being used in the `//Parse the fields` loop after the that base loop.

mikhailramalho commented on src/clang-c-frontend/clang_c_convert.h:
```diff
@@ -38,8 +38,12 @@ class StringLiteral;
 } // namespace clang
 
+extern std::string
+getFullyQualifiedName(const clang::QualType &, const clang::ASTContext &);
```
This doesn't need to be extern, since its definition is in the cpp file. 

Also, why do you need this?

kunjsong01 replied:
Done removed `extern`. 

>why do you need this?

Need to be used in clang-cpp-converter to get the parent_class_id

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -268,23 +278,58 @@ bool clang_cpp_convertert::get_struct_union_class_fields(
+    for(auto base : cxxrd->bases())
+    {
+      // `base` type is clang::CXXBaseSpecifier
+      base.getType().dump();
```
cleanup this.

kunjsong01 replied:
done

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -268,23 +278,58 @@ bool clang_cpp_convertert::get_struct_union_class_fields(
+    if(is_field_global_storage(field))
+      continue;
+
+    // [C++ annotation]: set parent in component's type
```
do we need to set this field?

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -807,8 +866,14 @@ bool clang_cpp_convertert::get_function_body(
-          log_error("Base class initializer is not supported in {}", __func__);
-          abort();
+          // TODO-split: add base class initializer
+          init->getInit()->dump();
```
cleanup this

kunjsong01 replied:
Done removed

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -807,8 +866,14 @@ bool clang_cpp_convertert::get_function_body(
+          // Add additional annotation for `this` parameter
+          initializer.derived_this_arg(
+            ftype.arguments().at(0).get("#identifier"));
+          initializer.base_ctor_derived(true);
```
why do we need to call derived_this_arg and base_ctor_derived?

I think the get_expr in the next line will overwrite anything set in this initializer object.

kunjsong01 replied:
>why do we need to call derived_this_arg and base_ctor_derived?

In short, the purpose is to generate typecasted `this` argument when calling base ctor in derived ctor, something like:
```cpp
 91 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 92
 93 Motorcycle (c:@S@Motorcycle@F@Motorcycle#):
 94         // 32
 95         FUNCTION_CALL:  Vehicle(( Vehicle *)this)  // <---- typecasted `this` argument
 96         // 33 file main.cpp line 13 function Motorcycle
 97         END_FUNCTION // Motorcycle
 98 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
"derived_this_arg" == symbol id for Motorcycle `this`
"base_ctor_derived" means "we are implicitly calling the base ctor in derivied ctor" 
(sorry about the wacky naming as I wan't able to give it a more intuitive and compact name :sweat_smile:) 

>I think the get_expr in the next line will overwrite anything set in this initializer object.

Correct... but we'll use them before the overwriting (i.e. the swapping) in `clang_cpp_convertert::get_constructor_call`. 

P.S. Combined with the new function `clang_cpp_convertert::gen_typecast_base_ctor_call`, I found these flags work pretty well when dealing with multiple inheritances (we'll see more test cases enabled for multiple inheritance in https://github.com/esbmc/esbmc/pull/884 when this PR is merged). Basically it's able to generate: 
```cpp
127 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
128 // given Rectangle class derives from Shape and PaintCost classes
129 Rectangle (c:@S@Rectangle@F@Rectangle#):
130         // 40
131         FUNCTION_CALL:  Shape(( Shape *)this)
132         // 41
133         FUNCTION_CALL:  PaintCost(( PaintCost *)this)
134         // 42 file main.cpp line 28 function Rectangle
135         END_FUNCTION // Rectangle
136 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
+  return false;
+}
+
+bool clang_cpp_convertert::get_access_from_decl(
```
do we actually need this? We are not doing any typecheck in the IR, so the access information is useless for us (clang already check if not code call violates access definitions).

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
+  std::string parent_class_name = getFullyQualifiedName(
+    ASTContext->getTagDeclType(cxxmdd->getParent()), *ASTContext);
+  std::string parent_class_id = tag_prefix + parent_class_name;
+  component_type.set("#member_name", parent_class_id);
```
do we need this information?

Also, instead of exporting getFullyQualifiedName, you should call `get_decl(cxxmdd->getParent())` and access its name.

I fail to see why we need this code tho.

kunjsong01 replied:
>Also, instead of exporting getFullyQualifiedName, you should call get_decl(cxxmdd->getParent()) and access its name.

well I thought of using `get_decl(cxxmdd->getParent(), new_expr)` ...  but it would appear to go through lots of recursions, seems to be an overkill and might be quite head scratching when we need to debug in the future (e.g. dealing with 25 frames in gdb). Obviously, here we just want the to get the id and annotate `member_name`. Using the getFullyQualifiedName function appears to be a neat way to do so.

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
+  // annotate name
+  std::string method_id, method_name;
+  clang_c_convertert::get_decl_name(*cxxmdd, method_name, method_id);
+  new_expr.name(method_id);
```
do we need this?

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
+  std::string access;
+  if(get_access_from_decl(cxxmdd, access))
+    return true;
+  new_expr.set("access", access);
```
do we need this?

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
+    return true;
+  new_expr.set("access", access);
+
+  // annotate base name
```
These two lines feel wrong...

base_name = unique identifier
pretty_name = name printed when we use --goto-functions-*, --symbol-table-*, --program-* and others

kunjsong01 replied:
I see... thanks very much for the explanations. I mistakes base_name for the parent class id. These annotations are added to match the symbol types generated by the old frontend. 


Now fixed: 
base_name = method id
pretty_name = method name

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
+  new_expr.pretty_name(parent_class_name);
+
+  // annotate inline
+  new_expr.set("is_inlined", cxxmdd->isInlined());
```
do we need this?

mikhailramalho commented on src/clang-cpp-frontend/clang_cpp_convert.cpp:
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
+
+  // We still need to add a trivial ctor/dtor as a `component` in class symbol's type
+  // remove "statement: skip" otherwise it won't be added
+  // TODO-split: Any more elegant way to do it?
```
Can you explain why we need this code?

kunjsong01 replied:
Preparation for vptr and vtable extensions. We need to make some implicit code for vptr later - sorry it may not be obvious for the time being.

mikhailramalho commented on src/util/irep.cpp:
```diff
@@ -538,6 +538,8 @@ const irep_idt irept::a_object_type = dstring("object_type");
 const irep_idt irept::a_cmt_size = dstring("#size");
 const irep_idt irept::a_cmt_type = dstring("#type");
 const irep_idt irept::a_type_id = dstring("typeid");
+const irep_idt irept::a_derived_this_arg = dstring("#derived_this_arg");
```
why do we need this?

mikhailramalho commented on src/util/irep.h:
```diff
@@ -264,6 +264,16 @@ class irept
     return get(a_pretty_name);
   }
 
+  inline const irep_idt &derived_this_arg() const
```
ditto

mikhailramalho commented on src/util/irep.h:
```diff
@@ -940,6 +950,16 @@ class irept
     set(a_pretty_name, val);
   }
 
+  inline void derived_this_arg(const irep_idt &val)
```
ditto

mikhailramalho commented on src/util/irep.h:
```diff
@@ -1232,6 +1252,8 @@ class irept
   static const irep_idt a_end_location, a_guard, a_label, a_lhs, a_location;
   static const irep_idt a_object_type, a_cmt_size, a_cmt, a_type_id;
   static const irep_idt a_cmt_type;
+  // annotations for typecasting derived derived class `this` to base class type
```
ditto

mikhailramalho reviewed (changes requested):
@kunjsong01 I added several comments, including a lot of nitpick. Remember to remove unnecessary code (calls to .dump()) and run clang-format for the next version).

Please, double-check all the new attributes we are setting and if they are actually needed.

Your patches are getting increasingly harder to review (that's why I asked you to explain some of your decisions). This is a good sign.

rafaelsamenezes commented on src/clang-c-frontend/clang_c_convert.cpp:
```diff
@@ -298,13 +303,22 @@ bool clang_c_convertert::get_struct_union_class(const clang::RecordDecl &rd)
   if(context.find_symbol(id) != nullptr)
     return false;
 
+  // TODO: Fix me when we have a test case using C++ unioin.
```
unioin =>  union

kunjsong01 replied:
done fixed.

rafaelsamenezes reviewed (changes requested):
Could you add more test-cases? For: static fields, nested inheritance, and pointer casts .

kunjsong01 committed:
[clang-cpp-converter] fixed more comments

kunjsong01 commented:
Hi @mikhailramalho , thanks a lot for the insightful reviews (and the explanations). I've fixed most of them. 

>do we need this? 

Good question.  
Honestly I also feel that I've added more annotations than possibly required. It would appear that adding the additional annotations might seem unnecessary, and the purpose is not quite obvious .

It's difficult to say for sure whether we need them or not. I just added these annotations to match the symbol table as much as I can. IIRC, some of them are needed to generate vtable and vptr. I may need them to generate implicit code in ctors/dtors for vptr stuff (this will become another PR after https://github.com/esbmc/esbmc/pull/884).

Apologies for not being able to give a concrete answer to this question. In order to speed up the develoment, I'm currently adopting a "learn-and-try" style development. As the development moves forward, it would become clearer. 

For the time being, I'd suggest let's leave it be and focus on pushing the development to pass more test cases (and yes we have more TCs passing in https://github.com/esbmc/esbmc/pull/884 :smiley:). But please be reassured that I'll purge these annotations after passing the virtual method test cases, and also please be assured that for any additional annotation to be added I'll try to confine them in a function so that it would be easy to indentify the redundant in the future.

kunjsong01 commented:
> Could you add more test-cases? For: static fields, nested inheritance, and pointer casts .

Hi @rafaelsamenezes thanks very much for the review. 
Good point. But this would be out of scope for this PR. I've got more test cases passing in the next PR https://github.com/esbmc/esbmc/pull/884 (which is currently gated by this PR) :smiley:. Currently, the focus of development is to pass the inheritance test suite in esbmc-cpp/inheritance.

mikhailramalho commented:
> Hi @mikhailramalho , thanks a lot for the insightful reviews (and the explanations). I've fixed most of them.
> 
> > do we need this?
> 
> Good question. Honestly I also feel that I've added more annotations than possibly required. It would appear that adding the additional annotations might seem unnecessary, and the purpose is not quite obvious .
> 
> It's difficult to say for sure whether we need them or not. I just added these annotations to match the symbol table as much as I can. IIRC, some of them are needed to generate vtable and vptr. I may need them to generate implicit code in ctors/dtors for vptr stuff (this will become another PR after #884).
> 
> Apologies for not being able to give a concrete answer to this question. In order to speed up the develoment, I'm currently adopting a "learn-and-try" style development. As the development moves forward, it would become clearer.
> 
> For the time being, I'd suggest let's leave it be and focus on pushing the development to pass more test cases (and yes we have more TCs passing in #884 smiley). But please be reassured that I'll purge these annotations after passing the virtual method test cases, and also please be assured that for any additional annotation to be added I'll try to confine them in a function so that it would be easy to indentify the redundant in the future.

Alright, given that this is not a critical part of esbmc, I agree to accept it for now.

But please, remember that: 
(1) esbmc is not well optimized for memory usage, so every extra annotation you add that is not necessary, more memory we are using unnecessarily.
(2) The more annotations you add, the more trouble you'll have when rewriting the frontend to use irep2 in the future. For instance, there is no access annotation in irep2. They are not even migrated in the migration code. I strongly believe that these are not necessary, and you might waste time migrating them to irep2 before getting a PR rejected.

My point is: be wary of your time :)

mikhailramalho commented:
Also, it's weird that the code style bot didn't complain about the unordered headers, is this option disabled @rafaelsamenezes? Can we enable it?

rafaelsamenezes commented:
> Also, it's weird that the code style bot didn't complain about the unordered headers

What do you mean by sorted? As in category or alphabetical?

For categories there are some regex that we can add: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#includecategories

Alphabetical is a matter of adding [SortIncludes](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#sortincludes) into .clang-format file.
